receiver/snmpreceiver/scraper.go

```diff
@@ -39,13 +64,345 @@ func newScraper(logger *zap.Logger, cfg *Config, settings component.ReceiverCrea
+		oid := configHelper.getResourceAttributeConfigOID(attributeName)
+		switch {
+		case prefix != "":
+			attributeValue = prefix + indexString
+		case oid != "":
+			indexedAttrKey := indexedAttributeKey{
+				parentOID: oid,
+				oidIndex:  indexString,
+			}
+			attributeValue = indexedResourceAttributeValues[indexedAttrKey]
+		default:
+			attributeValue = ""
+		}
+
+		// If no good attribute value could be found
+		if attributeValue == "" {
+			return nil, errors.New(errMsgResourceAttributeEmptyValue)
+		}
+
+		resourceAttributes[attributeName] = attributeValue
+	}
+
+	return resourceAttributes, nil
+}
+
+// scrapeIndexedAttributes retrieves all SNMP data from attribute (or resource attribute)
+// config column OIDs and stores the returned indexed data for later use by metrics
+func (s *snmpScraper) scrapeIndexedAttributes(
+	columnOIDs []string,
+	scraperErrors *scrapererror.ScrapeErrors,
+) map[indexedAttributeKey]string {
+	indexedAttributeValues := map[indexedAttributeKey]string{}
+
+	// If no OID resource attribute configs, nothing else to do
+	if len(columnOIDs) == 0 {
+		return indexedAttributeValues
+	}
+
+	// Retrieve all SNMP indexed data from column resource attribute OIDs
+	indexedData := s.client.GetIndexedData(columnOIDs, scraperErrors)
+
+	// For each piece of SNMP data, store the necessary info to help create resources later if needed
+	for _, data := range indexedData {
+		if err := indexedDataToAttribute(data, indexedAttributeValues); err != nil {
+			scraperErrors.AddPartial(1, fmt.Errorf(errMsgIndexedAttributeOIDProcessing, data.oid, data.parentOID, err))
+		}
+	}
+
+	return indexedAttributeValues
+}
+
+// indexedDataToAttribute provides a function which will take one piece of column OID SNMP indexed data
+// (for either an attribute or resource attribute) and stores it in a map for later use (keyed by both
+// {resource} attribute config column OID and OID index)
+func indexedDataToAttribute(
+	data SNMPData,
+	indexedAttributeValues map[indexedAttributeKey]string,
+) error {
+	// Get the string value of the SNMP data for the {resource} attribute value
+	var stringValue string
+	// Not explicitly checking these casts as this should be made safe in the client
+	switch data.valueType {
+	case notSupportedVal:
+		return fmt.Errorf(errMsgIndexedAttributesBadValueType, data.oid, data.parentOID)
+	case stringVal:
+		stringValue = data.value.(string)
+	case integerVal:
+		stringValue = strconv.FormatInt(data.value.(int64), 10)
+	case floatVal:
+		stringValue = strconv.FormatFloat(data.value.(float64), 'f', 2, 64)
+	}
+	// Store the {resource} attribute value in a map using the column OID and OID index associated
+	// as keys. This way we can match indexed metrics to this data through the {resource} attribute
+	// config and the indices of the individual metric values
+	indexString := strings.TrimPrefix(data.oid, data.parentOID)
+	attrKey := indexedAttributeKey{
+		parentOID: data.parentOID,
+		oidIndex:  indexString,
+	}
```

Comment:
This compound key ends up being a bit awkward because really there is a natural hierarchy of data here (column -> index).

It makes it more difficult than necessary to do simple checks later, on the number of indices in each column, which will be necessary to detect mismatches.

Seems like switching to `map[columnOID]map[indexString]string` would be better.

Also, is `parentOID` synonymous with `columnOID`? If so, can we standardize on `columnOID`? If not, what's the difference?

Reply:
I agree `map[string]map[string]string` type should be aliased. Probably just aliasing the inner is fine.

```go
type indexedAttributeValues map[string]string

// later
columnIndexes := map[string]indexedAttributeValues
```

Alias or not, it seems there would be a variety of error cases that we are not checking for, and this would make it much easier to do so. This can be a followup PR though.

Reply:
> @djaglowski switching to this could have usefulness....but only after also refactoring the receiver to some degree. right now as it is, the scraper is only processing a single piece of data at a time. this is because the client is given a number of column OIDs, and it returns a bunch of indexed data from all of those column OIDs.
>
> so as is, this combined key is works perfectly fine for this scraper design because again we're only accessing 1 piece of data at a time (the data does contain which column OID it came from) with no context to any sort of column OID grouping.
>
> i'll change this (mostly because i need to get this through this week so i'll do whatever it takes), but it's a change that isn't even relevant with the current design IMO.

Yes, that is the point! We are not doing any validation to ensure columns are of the same length. The reason we are not able to check for these error cases is because the current design makes it difficult - a relevant problem to resolve I would think.

Reply:
@djaglowski i'm just trying to say that the client would also have to be changed in how it returns data (maybe batches by column OID) in order for us to actually do anything with this particular change. either that or we'd have to scan through all returned data from the client to group them by column ID after grabbing from the client.

i'm not saying this is a bad change...i'm just questioning if this should have just been a "nice to have for future reasons" vs. a "you should do this"